streamalert/shared/config.py

```diff
@@ -284,8 +268,42 @@ def _validate_config(config):
                            for log_keys in
                            list(config[TopLevelConfigKeys.NORMALIZED_TYPES].values())):
                     raise ConfigError(
-                        'IOC key \'{}\' within IOC type \'{}\' must be defined for at least '
-                        'one log type in normalized types'.format(normalized_key, ioc_type)
+                        "IOC key '{}' within IOC type '{}' must be defined for at least "
+                        "one log type in normalized types".format(normalized_key, ioc_type)
                     )
 
-    # FIXME (derek.wang) write a configuration validator for lookuptables (new one)
+def _validate_sources(cluster_name, cluster_attrs, sources):
+    """Validates the sources for a cluster
+    Args:
+        data_sources (dict): The sources to validate
+    Raises:
+        ConfigError: If the validation fails
+    """
+    # Iterate over each defined source and make sure the required subkeys exist
+    data_sources = cluster_attrs['data_sources']
+    supported_sources = {'kinesis', 's3', 'sns', 'stream_alert_app'}
+    if not set(data_sources).issubset(supported_sources):
+        invalid_sources = set(data_sources) - supported_sources
+        raise ConfigError(
+            'The data sources for cluster {} contain invalid source entries: {}. '
+            'The following sources are supported: {}'.format(
+                cluster_name,
+                ', '.join("'{}'".format(source) for source in invalid_sources),
+                ', '.join("'{}'".format(source) for source in supported_sources)
+            )
+        )
+    for attrs in data_sources.values():
+        for entity, entity_attrs in attrs.items():
+            if TopLevelConfigKeys.LOGS not in entity_attrs:
```

Comment:
this could be removed if the comment above is taken into consideration. I think we're doing more work than necessary by nesting in a `logs` key